nixos/modules/services/x11/display-managers/default.nix

```diff
@@ -365,6 +373,63 @@ in
         StopWhenUnneeded = false;
       };
     };
+
+    # Create desktop files and scripts for starting sessions for WMs/DMs
+    # that do not have upstream session files (those defined using services.{display,desktop,window}Manager.session options).
+    services.xserver.displayManager.sessionPackages =
+      let
+        dms = filter (s: s.manage == "desktop") cfg.displayManager.session;
+        wms = filter (s: s.manage == "window") cfg.displayManager.session;
+
+        # Script responsible for starting the window manager and the desktop manager.
+        xsession = wm: dm: pkgs.writeScript "xsession" ''
+          #! ${pkgs.bash}/bin/bash
+
+          # Legacy session script used to construct .desktop files from
+          # `services.xserver.displayManager.session` entries. Called from
+          # `sessionWrapper`.
+
+          # Start the window manager.
+          ${wm.start}
+
+          # Start the desktop manager.
+          ${dm.start}
+
+          ${optionalString cfg.updateDbusEnvironment ''
+            ${lib.getBin pkgs.dbus}/bin/dbus-update-activation-environment --systemd --all
+          ''}
+
+          test -n "$waitPID" && wait "$waitPID"
+
+          ${config.systemd.package}/bin/systemctl --user stop graphical-session.target
+
+          exit 0
+        '';
+      in
+        # We will generate every possible pair of WM and DM.
+        crossLists
+          (dm: wm: let
+            sessionName = "${dm.name}${optionalString (wm.name != "none") ("+" + wm.name)}";
+            script = xsession dm wm;
+          in pkgs.writeTextFile {
+            name = "${sessionName}-xsession";
+            destination = "/share/xsessions/${sessionName}.desktop";
+            # Desktop Entry Specification:
+            # - https://standards.freedesktop.org/desktop-entry-spec/latest/
+            # - https://standards.freedesktop.org/desktop-entry-spec/latest/ar01s06.html
+            text = ''
+              [Desktop Entry]
+              Version=1.0
+              Type=XSession
+              TryExec=${script}
+              Exec=${script}
+              Name=${sessionName}
+              Comment=
```

Comment:
Would it be worth adding a `DesktopNames` key here?

Reply:
Hmm https://gitlab.freedesktop.org/xdg/xdg-specs/commit/729a8f2dde47694890d87f6e3f91527659f85834

Not sure what would the value be, though.

Reply:
How about an optional string `dm.name` upcased that excludes `none`?
```
${optionalString (dm.name != "none") ''
DesktopNames=${dm.name}
''}
```

can't remember if there's a capitalize function in nix.

Reply:
The purpose of `DesktopNames` is setting `XDG_CURRENT_DESKTOP` environment variable. Given that the variable is primarily used for filtering desktop files, it makes little sense to use our `DE_NAME+WM_NAME`, as nothing will know about them and use them. Well known desktop environments will use upstream session files.